dapp/request_limitation/throttling.go

```diff
@@ -0,0 +1,63 @@
+package request_limitation
+
+import (
+	"sync"
+	"time"
+)
+
+type Throttling struct {
+	concurrency    uint
+	coolDown       time.Duration
+	stack          chan func()
+	lock           sync.Mutex
+	maxQueue       uint
+	inWork         uint
+	queueFullError error
+}
+
+// create new throttling request limitation
+func NewThrottling(concurrency uint, coolDown time.Duration, maxQueue uint, queueFullError error) *Throttling {
+
+	t := &Throttling{
+		concurrency:    concurrency,
+		coolDown:       coolDown,
+		stack:          make(chan func(), maxQueue),
+		lock:           sync.Mutex{},
+		maxQueue:       maxQueue,
+		queueFullError: queueFullError,
+	}
+
+	go func() {
+		for {
+			t.lock.Lock()
+			if t.inWork >= t.concurrency {
+				t.lock.Unlock()
+				continue
```

Comment:
This continue here would re-start the infinite for loop if t.inWork >= t.concurrency and since both t.inWork and t.concurrency are not being modified within the if condition, it would loop infinitely unless an external force changes either t.inWork or t.concurrency.

Even if an external force is quick enough to modify t.inWork or t.concurrency, the continue statement would likely trigger many iterations of the for loop before the external force changes one of the two values.

Adding a time.Sleep(time.Second) might decrease the performance impact if an infinite loop is triggered in an edge case condition.

Reply:
Fixed